src/BatteryStateOfChargeService.cpp

```diff
@@ -16,21 +20,57 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    double diffOfTimeMSec = prevTime->msecsTo(myTime);
+    double diffOfTime = diffOfTimeMSec/3600000;
+    double avgCurrent = (myCurrent+prevCurrent)/2;
+
+    double changeInAmpHour = avgCurrent * diffOfTime;
+    ampHourTotal += changeInAmpHour;
+    //if(ampHourTotal > BATTERY_AMP_HOUR_CAPACITY){ampHourTotal = BATTERY_AMP_HOUR_CAPACITY;}
+    ampHourUsed = BATTERY_AMP_HOUR_CAPACITY- ampHourTotal;
+    return ampHourUsed;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
-    return false;
+    if(myCurrent>=0)
+    {
+        return false;
+    }
+    return true;
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+    double ampHoursLeft;
+    if(isCharging())
+    {
+        ampHoursLeft = BATTERY_AMP_HOUR_CAPACITY - ampHourTotal;
+    }
+    else
+    {
+        ampHoursLeft = ampHourTotal;
+    }
+    int current = myCurrent;
+    if(myCurrent < 0)
+    {
+        current *= -1;
+    }
+    double msecLeft = (ampHoursLeft * 3600000) / current;
+    QTime QTimeLeft = QTime(0,0,0,0);
+    QTimeLeft = QTimeLeft.addMSecs(msecLeft);
+    return QTimeLeft;
 }
 
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
-    Q_UNUSED(batteryData);
-    // Update your variables here.
+    if(myTime.isNull())
+    {
+        prevTime = new QTime(0,0,0,0);
+    }
+    myVoltage = batteryData.voltage;
+    prevCurrent = myCurrent;
+    myCurrent = batteryData.current;
+    prevTime = new QTime(myTime.hour(),myTime.minute(),myTime.second(),myTime.msec());
```

Comment:
Everytime you use `new` you are allocating dynamic memory, which also means you are in charge of deallocating the same memory using `delete`. In this case, you have a memory leak because you are never deallocating this memory.